bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java

```diff
@@ -298,8 +302,24 @@ public ResultScanner getScanner(final Scan scan) throws IOException {
     LOG.trace("getScanner(Scan)");
     Span span = TRACER.spanBuilder("BigtableTable.scan").startSpan();
     try (Scope scope = TRACER.withSpan(span)) {
+      ResultScanner scanner;
+      if (scan.getCaching() == -1) {
+        scanner = clientWrapper.readRows(hbaseAdapter.adapt(scan));
+      } else {
+        final Query request = hbaseAdapter.adapt(scan);
+        final RequestContext requestContext =
+            RequestContext.create("ProjectId", "InstanceId", "AppProfile");
+        int requestedPageSize = (int) request.toProto(requestContext).getRowsLimit();
+        if (requestedPageSize > scan.getCaching()) {
+          requestedPageSize = scan.getCaching();
+        }
+        if (requestedPageSize <= 0) {
+          requestedPageSize = PAGE_SIZE;
+        }
 
-      final ResultScanner scanner = clientWrapper.readRows(hbaseAdapter.adapt(scan));
+        Query.QueryPaginator paginator = request.createPaginator(requestedPageSize);
```

Comment:
requestedPageSize should be scan.getCaching(). The only thing we need to verify is scan.getCaching() is > 0.

Reply:
Done

Reply:
Done